Improvements to the evaluate function focusing on optimizing derived predicates and existential preconditions resolution

Rezenders opened the pull request:
This PR is a draft to solve issue #328. 

I implemented the algorithm described by @tobiaswjohn for most of the Plansys2 Node types.

This version is much faster than the previous version in the main branch. It takes around 37643ms to solve all derived predicates in [this PDDL formulation](https://github.com/Rezenders/ros2_planning_system/blob/new_evaluation/plansys2_problem_expert/test/pddl/domain_derived.pddl), whereas it wouldn't finish with the previous version of the algorithm.

The evaluate function could be further improved by:
-  by feeding the already assigned variables to the eval function
- with an algorithm that checks the type of the derived predicate parameters and finds out the possible types for the variables in the derived predicate preconditions, which could reduce the search space significantly in certain cases

I still need to adjust the other packages to make it compatible with the new evaluation and problem expert.

I didn't adjust the evaluation for anything related to `functions`. If you think this approach would be mergeable when complete, I can adjust it to handle `functions`. But right now, I don't need it for my project, so I will focus on finishing what I need first.

Rezenders committed:
add == operator and hash to Instance and Predicate

Rezenders committed:
change checkNodeEquality to consider predicates w/ variables

Rezenders committed:
add checkParamTypeEquivalence

Rezenders committed:
⚡ new evaluate function (issue #328)

Rezenders committed:
🔥 rm cart_product and replace_children_param

Rezenders commented:
Possible improvements:
- Move [StateVec](https://github.com/PlanSys2/ros2_planning_system/blob/b82cc6fe5d577e481b0aa01b6b779f94d33b9ebe/plansys2_executor/include/plansys2_executor/bt_builder_plugins/stn_bt_builder.hpp#L38C1-L42C3) to the `plansys2_problem_expert` (or plansys2_core) package and add a `StateVect getState()` method to problemExpert. Then use the StateVec as an input to the evaluation function instead of passing `instances`, `predicates` and `functions`
- Do we really need to get the state and pass it to the evaluation? Can't we just use the problemExperClient? It would make things much simpler

Rezenders committed:
[plansys2_executor] use instances in evaluation

Rezenders committed:
[evaluate] return true when apply and negate are true

Rezenders committed:
adjust plansys2_executor tests to new evaluate

Rezenders committed:
🎨 fix code style

Rezenders committed:
fixes to Expression and Ground getTree

Rezenders committed:
fixes to evaluate and helper functions

Rezenders commented:
When the evaluate function is called with `apply=true` and `use_state=true`, the derived predicates are not recalculated after the state is modified. Check: https://github.com/Rezenders/ros2_planning_system/blob/b6c996cc2be1c993c8cbfb80c24e09e758599998/plansys2_problem_expert/src/plansys2_problem_expert/Utils.cpp#L105-L113 

This is a crucial issue when generating the plan graph. 
To solve this problem, I propose creating a new `State` class containing the instances, functions, predicates, inferred_predicates, and derived predicates. 

```
class State
  instances
  predicates
  inferred_predicates
  derived_predicates
  functions
```

Then:
1. pass an instance of `State` to the evaluate function instead of passing the instances, functions, and predicates
2. after an apply operation, clear the `inferred_predicates` and use the `derived_predicates` to recalculate them
3. add a `State` msg to plansys2_msgs
4. add a `getState` method to the ProblemExpert
5. add a get_state service to the ProblemExpertNode
6. add a `getState` method to the ProblemExpertClient


```
evaluate(tree, problem_client, state, apply, use_state, node_id, negate):
...
case Predicate:
 if(apply):
   if(use_state):
     apply_operation(state.predicates)
     state.inferred_predicates.clear()
     state = all_derived(state)
...
```

Rezenders committed:
[simple_bt_builder] add check_requirement

Rezenders committed:
🔥 rm comments

Rezenders committed:
🔥 rm code commited by mistake

Rezenders commented:
> To solve this problem, I propose creating a new State class containing the instances, functions, predicates, inferred_predicates, and derived predicates.

Actually, I think the State class shouldn't contain `derived_predicates`. The `derived_predicates` would only be needed on apply operations. Maybe it is better to separate the `apply` function from the `evaluate` function, and pass the `derived_predicates` to the apply function, e.g., `apply(tree, problem_client, state, derived_predicates)`. It would make the evaluate function more readable. However, I don't know how much code would be repeated in both.

```
class State
  instances
  predicates
  inferred_predicates
  functions
```

Rezenders committed:
getDerivedPredicates return vector<Derived>

Rezenders committed:
⚡ std::move(param_values) and evaluate return

Rezenders committed:
⚡ improve groundPredicate performance

Rezenders committed:
🚚 move solveAllDerived, solveDerived and groundPredicates to utils

Rezenders committed:
⚡ improve mergeParamsValuesVector performance

Rezenders committed:
✨ add State class to plansys2_core

Rezenders committed:
add state_ : State to ProblemExpert

Rezenders committed:
[plansys2_msgs] add State.msg and GetState.srv

Rezenders committed:
[plansys2_core] update State

Rezenders committed:
⚡ use state everywhere and move solveDerived to Utils

Rezenders committed:
update tests

Rezenders committed:
commenting out stn_bt_builder in CMakeLists

Rezenders committed:
[plansys2_executor] use state to evaluate/apply

Rezenders committed:
[plansys2_terminal] adjust to new getDerivedPredicates

Rezenders committed:
[#331]apply at_end_effects in get_node_contradict

Rezenders commented:
> When the evaluate function is called with apply=true and use_state=true, the derived predicates are not recalculated after the state is modified. 

Solved

I am facing a new problem now. The BT generator algorithm [simple_bt_builder] calls the apply function many times, triggering the resolution of all derived predicates every time the apply function is called. Even though the resolution of derived predicates is much faster now, it can still take up to a few seconds to solve hard problems. Thus, the BT generation is taking quite a long time.

I will check whether the simple_bt_builder algorithm can be optimized with some early returns and work on implementing the optimizations described in #328 

PS: I believe there might be some unnecessary copying of the `state` variable in some places. I will check if I can optimize the code by moving the data instead of deep copying

Rezenders committed:
⚡ only solveDerived once when applying effects

fmrico commented:
Sorry for the conflicts, @Rezenders . There were some things to fix

Please, merge, or better, rebase

Rezenders commented:
Ideas to improve BT generator algorithm:
1. cache the resulting state of applying an action to a State, e.g., `cache[state, action] = apply(action.effects, state)`. This would probably require hashing a State individual which I believe has complexity `O(size(State))`
2. early return in `get_node_satisfy` when the first satisfying node is found. Here: https://github.com/PlanSys2/ros2_planning_system/blob/ab59c45af515e32e0570ec33ca24c6d5311787bd/plansys2_executor/src/plansys2_executor/bt_builder_plugins/simple_bt_builder.cpp#L110-L112 
Also here: https://github.com/PlanSys2/ros2_planning_system/blob/ab59c45af515e32e0570ec33ca24c6d5311787bd/plansys2_executor/src/plansys2_executor/bt_builder_plugins/simple_bt_builder.cpp#L209 
Does this make sense?
3. Check if the `get_node_satisfy` and `get_node_contradict` loops can be merged to avoid unnecessary loops

Rezenders commented:
> Sorry for the conflicts, @Rezenders . There were some things to fix
> 
> Please, merge, or better, rebase

Ok, no worries :)

Rezenders committed:
add hash and == operator to State

Rezenders committed:
[simple_bt_builder] add test for domain with derived

Rezenders committed:
problem_expert: add derive argument to apply

Rezenders committed:
[simple_bt_builder] add action_state_cache

Rezenders commented:
Adding an <action, state> cache reduced the time for creating a graph from around 145 seconds to 13 seconds in the following test:
https://github.com/Rezenders/ros2_planning_system/blob/da7feb9d1ff92cdc5b25f9c693f7b29ec8972a22/plansys2_executor/test/unit/simple_btbuilder_tests.cpp#L958

Rezenders committed:
add Graph class

Rezenders committed:
add getDerivedPredicateGraph

Rezenders committed:
solve derived predicates with graph

Rezenders committed:
add checkActionEquality

Rezenders committed:
add Action class

Rezenders committed:
updates to ActionVariant

Rezenders committed:
add ActionVariant and backtraverse to Graph

Rezenders committed:
add tests to Graph and State

Rezenders committed:
solve derived predicates with graph algorithm

Rezenders committed:
create cpp for plansy2_core headers

Rezenders committed:
don´t solve derived predicates with getState

Rezenders committed:
ActionVariant moved to plansys2_core

Rezenders committed:
move BTBuilder to bt_builder namespace

This was required due to conflicts between the Graph structure defined
in BTBuilder and the new Graph class in plansys2_core

Rezenders committed:
early return to get_node_satisfy

Rezenders committed:
simple_btbuilder: prune & solve derived predicates

Rezenders committed:
🎨 fix code style

Rezenders committed:
add State.hasPredicate(str)

Rezenders committed:
increase timeout when calling validate domain service

Rezenders committed:
fix small bugs

Rezenders committed:
solving derived predicates in BT action nodes

Rezenders commented:
The derived predicates graph algorithm described in #328 is implemented. The time for creating the plan graph is now around 1.3 seconds.
I also changed the BT nodes to solve the derived predicates when needed. The whole planning->plan graph generation->plan execution pipeline is now working. I added multiple unit tests and I tested it in simulation with SUAVE.

Some improvements can still be made but I will work on it later. For example, adding a <action, state> cache in the BT nodes.

I am sorry that the PR got super long, it will probably be annoying to review it.

fmrico commented:
Hi @Rezenders 

We have to retake this work. I think it is useful. I see two problems:
1. Current file conflicts.
2. It is super long, as you said.

How easy would it be for you to, based on this draft PR, start doing small and very focused PR to obtain the final result?

Rezenders commented:
> Hi @Rezenders
> 
> We have to retake this work. I think it is useful. I see two problems:
> 
> 1. Current file conflicts.
> 2. It is super long, as you said.
> 
> How easy would it be for you to, based on this draft PR, start doing small and very focused PR to obtain the final result?

Hi @fmrico, sorry for the delay. I will start working on this in the first week of June.
I agree on making small PRs. I will focus on that.

Rezenders committed:
fix getExpr - regex

Rezenders committed:
getExpr fix parsing for words with hyphen

fmrico commented:
> > Hi @Rezenders
> > We have to retake this work. I think it is useful. I see two problems:
> > 
> > 1. Current file conflicts.
> > 2. It is super long, as you said.
> > 
> > How easy would it be for you to, based on this draft PR, start doing small and very focused PR to obtain the final result?
> 
> Hi @fmrico, sorry for the delay. I will start working on this in the first week of June. I agree on making small PRs. I will focus on that.

Perfect. I will try to be reactive to this. May we close this PR and start with the new ones?

Rezenders committed:
add parser::pddl::removeOperatorBeforeParenthesis

Rezenders committed:
fix bug in getSubExpr

Rezenders committed:
getNodeType works with parameters

Rezenders committed:
fromString works for Parameter

Rezenders committed:
fix code style

Rezenders committed:
🚧 refactor Graph class

Rezenders committed:
WIP fixing bugs in graph etc

Rezenders committed:
🚧 graph + tarjan's + dfs + nodevariant file

Rezenders committed:
compare_str_case_insensitive

Rezenders committed:
add export_graph to plansys2_tools

Rezenders committed:
Graph -> DerivedResolutionGraph

Rezenders committed:
add state.getDerivedPredicatesSCCs

Rezenders committed:
DerivedGraph -> DerivedResolutionGraph

Rezenders committed:
add evaluateSCC

Rezenders committed:
ignore param type in checkParamEquality

Rezenders committed:
tests passing for current suave formulation

Rezenders committed:
fix suave_bt_execution_test

Rezenders committed:
fix Exists parse

Rezenders committed:
fix toStringExists and toString for parameter type

Rezenders committed:
add tests for toString and Exsits parse

Rezenders committed:
fix export_derived_graph

Rezenders committed:
fix negate in evaluate

Rezenders committed:
Fix + imp State class

Rezenders committed:
use new State interfaces

Rezenders committed:
improved Derived hashing

Rezenders committed:
improve addInferredPredicate

Rezenders committed:
improve groundPredicate performance

Rezenders committed:
improve hashing and == operators

Rezenders committed:
checkNodeEquality rm case insensitive comparison

Rezenders committed:
remove prints

Rezenders committed:
improve hashing and ==

Rezenders committed:
checkActionEquality case sensitive

Rezenders committed:
more test cases utils.cpp

Rezenders committed:
performance improvments to simplebtbuilder

Rezenders committed:
remove unused variable

Rezenders committed:
adjust tests pddl_parser

Rezenders committed:
fix evaluation negation

Rezenders committed:
add sequential bt builder

Rezenders committed:
change default BT for SequentialBTBuilder

Rezenders committed:
Ground.cpp fix parsing of constant

Rezenders committed:
fix Ground PDDLPrint for constant

Rezenders committed:
add plansys2_msgs/srv/AffectNodes.srv

Rezenders committed:
add addPredicates

Rezenders committed:
add plansys2_msgs/srv/UpdateNodes.srv

Rezenders committed:
add removePredicates and updatePredicates

Rezenders committed:
wip
